app/plugins.go

```diff
@@ -84,3 +90,129 @@ func InitPlugins() {
+	}
+
+	filenames, err := utils.ExtractTarGz(pluginFile, Srv.PluginEnv.SearchPath())
+	if err != nil {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.extract.app_error", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	if len(filenames) == 0 {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.no_files.app_error", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	splitPath := strings.Split(filenames[0], string(os.PathSeparator))
+
+	if len(splitPath) == 0 {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.bad_path.app_error", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	manifestDir := filepath.Join(Srv.PluginEnv.SearchPath(), splitPath[0])
+
+	manifest, _, err := mmplugin.FindManifest(manifestDir)
+	if err != nil {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.manifest.app_error", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	// Should add manifest validation and error handling here
+
+	err = Srv.PluginEnv.ActivatePlugin(manifest.Id)
+	if err != nil {
+		return nil, model.NewAppError("UnpackAndActivatePlugin", "app.plugin.activate.app_error", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	return manifest, nil
+}
+
+func GetActivePluginManifests() ([]*mmplugin.Manifest, *model.AppError) {
+	if Srv.PluginEnv == nil || !*utils.Cfg.PluginSettings.Enable {
+		return nil, model.NewAppError("GetActivePluginManifests", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
+	}
+
+	plugins, err := Srv.PluginEnv.Plugins()
+	if err != nil {
+		return nil, model.NewAppError("GetActivePluginManifests", "app.plugin.get_plugins.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	manifests := make([]*mmplugin.Manifest, len(plugins))
+	for i, plugin := range plugins {
+		manifests[i] = plugin.Manifest
+	}
+
+	return manifests, nil
+}
+
+func RemovePlugin(id string) *model.AppError {
+	if Srv.PluginEnv == nil || !*utils.Cfg.PluginSettings.Enable {
+		return model.NewAppError("RemovePlugin", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
+	}
+
+	err := Srv.PluginEnv.DeactivatePlugin(id)
+	if err != nil {
+		return model.NewAppError("RemovePlugin", "app.plugin.deactivate.app_error", nil, err.Error(), http.StatusBadRequest)
+	}
+
+	err = os.RemoveAll(filepath.Join(Srv.PluginEnv.SearchPath(), id))
+	if err != nil {
+		return model.NewAppError("RemovePlugin", "app.plugin.remove.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return nil
+}
+
+// Temporary WIP function/type for experimental webapp plugins
+type ClientConfigPlugin struct {
+	Id         string `json:"id"`
+	BundlePath string `json:"bundle_path"`
+}
+
+func GetPluginsForClientConfig() string {
+	if Srv.PluginEnv == nil || !*utils.Cfg.PluginSettings.Enable {
+		return ""
+	}
+
+	plugins, err := Srv.PluginEnv.Plugins()
```

Comment:
This also returns plugins with broken manifests so we can do things like show useful errors in the eventual UI.

So I believe the `plugin.Manifest.Webapp` expression below could panic.

Reply:
Ok cool, will update to handle that